### PR TITLE
feat!: Refactor/simplify configuration code

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>

* Remove the ability to hot-reload config. Users should never rely on this. Some config changes required restart.
* Remove the risk of having incorrect config loaded briefly before start-up.

### CODE REVIEW

File: config/controller.go [hunk 1]
```diff
@@ -71,63 +51,14 @@ func (cc *controller) parseConfigMap(cm *apiv1.ConfigMap) (interface{}, error) {
...
-					}
-				}
-			},
-		})
-	controller.Run(stopCh)
-	log.Info("Watching config map updates")
-}
-
-func (cc *controller) Get(ctx context.Context) (interface{}, error) {
+func (cc *controller) Get(ctx context.Context, config interface{}) error {
```
Comment (terrytangyuan) [inline config/controller.go:57]: Should we re-use the configmap informer?

**Review by terrytangyuan — APPROVED** [review]

File: server/apiserver/argoserver.go [hunk 2]
```diff
@@ -376,15 +376,6 @@ func mustRegisterGWHandler(register registerFunc, ctx context.Context, mux *runt
 	}
 }
 
-// Unlike the controller, the server creates object based on the config map at init time, and will not pick-up on
-// changes unless we restart.
-// Instead of opting to re-write the server, instead we'll just listen for any old change and restart.
-func (as *argoServer) restartOnConfigChange(interface{}) error {
```
Comment (sarabala1979) [inline server/apiserver/argoserver.go:382]: Will this change the existing behavior of argo server?  Argo server will not restart if there is any config change. is it right?

Comment (alexec) [reply server/apiserver/argoserver.go:382]: Correct. This PR should be marked as a breaking change.

**Review by sarabala1979 — APPROVED** [review]

**Comment by alexec** [comment]
TestFunctionalSuite/TestEventOnNodeFailSentAsPod fails

**Pull request merged**